### ISSUE TITLE
Audio record speed is slightly wrong

### ISSUE THREAD

**Issue by DuncanStone** [naudio/NAudio#839]
Using WaveIn (also tried WaveInEvent) I am recording sound from the microphone. The playback speed I am getting when playing the resulting sound file is slightly faster than reality. If I record a metronome I am getting a slow desync. When I record 30 seconds of audio it writes a 28 second file. This causes videos I am muxing the sound onto to have a very noticeable a/v desync after about 10 seconds.

I was having this issue with MediaCapture so I switched to NAudio. NAudio is better by far, but still not right. MediaCapture was getting about 25 seconds of audio from 30 of recording.

I am open to the idea that I might have a hardware issue or be implementing the capture wrong (I used the example on the readme) given that the issue affected 2 different capture libararies. Nevertheless if I use the windows built in voice recorder app, the audio sync is spot on, so I am not sure what is going on.

Is anyone able to shed any light on this?

**Comment by markheath** [comment]
Not something I've come across before. Possibly some audio is being dropped? Does the same happen if you capture with say Wasapi? Also, might be worth sharing the code in your event handler - you need to handle the audio quickly and make sure you properly observe the bytes recorded property.

**Comment by DuncanStone** [comment]
I am using the same code from the wav example 
`waveIn.DataAvailable += (s, a) =>
{
	writer.Write(a.Buffer, 0, a.BytesRecorded);
	if (writer.Position > waveIn.WaveFormat.AverageBytesPerSecond * 35)
	{
		waveIn.StopRecording();
	}
};`

If the main UI thread is lagging could that cause audio to be dropped? What would the main resource be that could cause this? Probably a lot of disk access would be an issue. I am writing video frames to disk at the same time. Should I create a dedicated thread for the event handler?

**Comment by DuncanStone** [comment]
Sorry about the delay, I replied to this and somehow didn't click the comment button...

The code from the example I copied forthe event handler was:

    waveIn.DataAvailable += (s, a) =>
    {
    	writer.Write(a.Buffer, 0, a.BytesRecorded);
    	if (writer.Position > waveIn.WaveFormat.AverageBytesPerSecond * 35)
    	{
    		waveIn.StopRecording();
    	}
    };

It is quite possible that audio is being dropped, the system is recording video frames at the same time and is under some heavy resource use. What should I look for/change to make sure that no samples get dropped? Would it drop frames in the code above if the write operation takes too long? Do I need to implement a buffer swapping system or something?

**Comment by DuncanStone** [comment]
So I tried another test using instead the following code

    mciSendString("open new Type waveaudio Alias recsound", "", 0, 0);
    mciSendString("record recsound", "", 0, 0);
    ...
    mciSendString($"save recsound c:\\Temp\\test.wav", "", 0, 0);
    mciSendString("close recsound ", "", 0, 0);

and got the same result. I am pretty sure the fact that it is happening with all the recording systems I have tried means that this is not an issue with NAudio and I will understand if you want to close this report based on that.

Having said that I know there are some real experts here that might be able to shed some light on the issue. I am at a complete loss. There are actually two different places in my code that call the same audio record function, one of them is affected and the other isn't. I have gone through the code over and over however and I can find no substantial differences that could cause this. Does anyone have any idea what I should be looking for? CPU, memory and GPU usage are all within reasonable values during the recording process.
